spring/src/test/java/io/cloudevents/spring/webflux/WebFluxRestControllerTests.java

```diff
@@ -0,0 +1,201 @@
+		assertThat(headers).containsKey("ce-source");
+		assertThat(headers).containsKey("ce-type");
+
+		// assertThat(headers.getFirst("ce-id")).isNotEqualTo("12345");
+		assertThat(headers.getFirst("ce-type")).isEqualTo("io.spring.event.Foo");
+		assertThat(headers.getFirst("ce-source")).isEqualTo("https://spring.io/foos");
+
+	}
+
+	@Test
+	void structuredRequestResponseEvents() {
+
+		ResponseEntity<String> response = rest
+				.exchange(RequestEntity.post(URI.create("http://localhost:" + port + "/event")) //
+						.contentType(new MediaType("application", "cloudevents+json")) //
+						.body("{" //
+								+ "\"id\":\"12345\"," //
+								+ "\"specversion\":\"1.0\"," //
+								+ "\"type\":\"io.spring.event\"," //
+								+ "\"source\":\"https://spring.io/events\"," //
+								+ "\"data\":{\"value\":\"Dave\"}}"),
+						String.class);
+
+		assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
+		assertThat(response.getBody()).isEqualTo("{\"value\":\"Dave\"}");
+
+		HttpHeaders headers = response.getHeaders();
+
+		assertThat(headers).containsKey("ce-id");
+		assertThat(headers).containsKey("ce-source");
+		assertThat(headers).containsKey("ce-type");
+
+		// assertThat(headers.getFirst("ce-id")).isNotEqualTo("12345");
+		assertThat(headers.getFirst("ce-type")).isEqualTo("io.spring.event.Foo");
+		assertThat(headers.getFirst("ce-source")).isEqualTo("https://spring.io/foos");
+
+	}
+
+	@Test
+	void requestResponseEvents() {
+
+		ResponseEntity<String> response = rest
+				.exchange(RequestEntity.post(URI.create("http://localhost:" + port + "/event")) //
+						.header("ce-id", "12345") //
+						.header("ce-specversion", "1.0") //
+						.header("ce-type", "io.spring.event") //
+						.header("ce-source", "https://spring.io/events") //
+						.contentType(MediaType.APPLICATION_JSON) //
+						.body("{\"value\":\"Dave\"}"), String.class);
+
+		assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
+		assertThat(response.getBody()).isEqualTo("{\"value\":\"Dave\"}");
+
+		HttpHeaders headers = response.getHeaders();
+
+		assertThat(headers).containsKey("ce-id");
+		assertThat(headers).containsKey("ce-source");
+		assertThat(headers).containsKey("ce-type");
+
+		// assertThat(headers.getFirst("ce-id")).isNotEqualTo("12345");
+		assertThat(headers.getFirst("ce-type")).isEqualTo("io.spring.event.Foo");
+		assertThat(headers.getFirst("ce-source")).isEqualTo("https://spring.io/foos");
+
+	}
+
+	@SpringBootApplication
+	@RestController
+	static class TestApplication {
+
+		@PostMapping("/")
+		public ResponseEntity<Foo> echo(@RequestBody Foo foo, @RequestHeader HttpHeaders headers) {
+			CloudEvent attributes = CloudEventHttpUtils.fromHttp(headers).withId(UUID.randomUUID().toString())
+					.withSource(URI.create("https://spring.io/foos")).withType("io.spring.event.Foo").build();
+			HttpHeaders outgoing = CloudEventHttpUtils.toHttp(attributes);
+			return ResponseEntity.ok().headers(outgoing).body(foo);
```

Comment:
Same concern as before, I may misunderstood the logic of the test, but shouldn't you test here that the response cloudevent is properly written? Why are you manually performing this conversion?

Reply:
Because as a user I want to consume a `Foo` (POJO) and `CloudEvent` is not appropriate for that (as we discussed yesterday).

Reply:
Ok but then what happens if I send a structured event to `/`? I won't be able to parse it right? That's why I guess here you want to accept a `CloudEvent` and then with `PojoCloudEventDataMapper` extract `Foo` from it

Reply:
As a user, I don't really care if the CloudEvent is wrapped in a json (structured mode) or smashed in the headers (binary mode), I just want that the framework gives me the entity in a uniform way, then it's up to me to figure out extracting `Foo`. Then, as i mentioned yesterday, in next iterations we might improve it adding an annotation to extract `Foo` straight from `CloudEventData` or something like that.

Reply:
And it is (transparent), if you consume `CloudEvent`, as is show by the tests here. But if you consume `Foo`, currently the best option is to have Spring do the conversion; and then you have to deal explicitly with the headers. What might be useful is to have support for `CloudEventContext` as a pseudonym for `HttpHeaders`, but that's kind of an enhancement on top of what is here.

Reply:
Ok but in that case, the user explicitly accepts to consume only binary CloudEvents. Does this makes sense? Do we have use cases where the user wants to only explicitly accept binary CloudEvents? My question is: do we have a use case where a user would developer an endpoint like this one?

Reply:
By spec every implementation (senders and receivers) SHOULD support both binary and structured: https://github.com/cloudevents/spec/blob/v1.0/http-protocol-binding.md#13-content-modes

The way I see it is that in any eventual use case the user wants to interact with CloudEvevents, he/she must be able to accept both, otherwise its implementation won't be 100% compliant, causing problems when 2 systems have to communicate.

Reply:
> Do we have a use case where a user would developer an endpoint like this one?

Yes. It's quite possible to implement the complementary endpoint that accepts a `Foo` from a structured request body. I think I even did that in the original #293 PR. We can definitely refine this in the future, but the basic building blocks from the Spring side are already in place (content negotiation and deserialization).

Reply:
> The way I see it is that in any eventual use case the user wants to interact with CloudEvents, he/she must be able to accept both,

That's explicitly what the features in this PR supports directly (per the integration test example).

Reply:
> It's quite possible to implement the complementary endpoint that accepts a Foo from a structured request body

Why the user should define a complementary endpoint? Isn't that like exposing the difference between binary and structured, forcing the user to expose 2 endpoints for receiving the same entity in 2 different "encodings"?

I think I explained myself badly: I would love to give the user the ability to extract `Foo` from the endpoint signature just like you did in this test, but without exposing the problem binary/structured. Unless I miss something, my understanding here is that the user, to extract that `Foo`, needs to have 2 endpoints one for binary one for structured, am I right?

Reply:
Yes, I believe so. We can maybe try and fix that later, but for now that is how it is (and I for one don't think that's too terrible).

Reply:
TBH I think it's a double edged sword, but ok let's expose this stuff now and let's be explicit in the javadoc that this won't work with the structured mode, ok?

Reply:
I'm not sure that's accurate. It works fine in structured mode, but you have to be explicit about that in your handler. I'm not sure the Javadoc for the library jar can capture that adequately. What about the README? Or the integration tests?

Reply:
BTW the javadocs for `HttpUtils` already has this:

> The input headers must represent a valid event in "binary" form, i.e. it must have headers "ce-id", "ce-specversion" etc.

So that's pretty explicitly referring to binary mode in CE terms.